Sources/PackageGraph/PackageGraph+Loading.swift

```diff
@@ -523,34 +522,74 @@ private func createResolvedPackages(
+    if let rootPkg = rootPkgID {
+        var pkgStack = [PackageIdentity]()
+        walkPkgTreeAndGetModuleAliases(for: rootPkg, with: packagesByIdentity, using: &pkgStack, output: &result)
+    }
+    return result
+}
+
+// Walk a package dependency tree and set the module aliases for targets in each package.
+// If multiple aliases are specified in upstream packages, aliases specified most downstream
+// will be used.
+private func walkPkgTreeAndGetModuleAliases(for pkgID: PackageIdentity,
+                                             with packagesByIdentity: [PackageIdentity: ResolvedPackageBuilder],
+                                             using pkgStack: inout [PackageIdentity],
+                                             output result: inout [PackageIdentity: [String: String]]) {
+    // Get the builder first
+    if let builder = packagesByIdentity[pkgID] {
+        // Add the pkgID to a stack used to keep track of
+        // multiple aliases specified in the package chain
+        pkgStack.append(pkgID)
+        builder.package.targets.forEach { target in
+            target.dependencies.forEach { dep in
+                // Check if a dependency for this target has module aliases specified
+                if case let .product(prodRef, _) = dep {
+                    if let prodPkg = prodRef.package {
+                        if let prodModuleAliases = prodRef.moduleAliases {
+                            for (depName, depAlias) in prodModuleAliases {
+                                let prodPkgID = PackageIdentity.plain(prodPkg)
+                                // Add the specified alias and the dependency package to a map
+                                result[prodPkgID, default: [:]][depName] = depAlias
+                            }
+                        }
                     }
                 }
             }
+            
+            // If multiple aliases are specified in the package chain,
+            // use the ones specified most downstream to override
+            // upstream targets
+            for trackedPkg in pkgStack {
+                if let entry = result[trackedPkg],
+                    let aliasToOverride = entry[target.name],
+                    pkgID != trackedPkg {
+                    result[pkgID, default: [:]][target.name] = aliasToOverride
+                    break
+                }
+            }
         }
-    }}}
-    
-    var pkgToAliasesMap = [String: [String: String]]()
-    nameToAliasAndPkgMap.forEach { (keyName, aliasAndPkgList) in
-        aliasAndPkgList.forEach { (elemAlias, elemPkg) in
-            // FIXME: handle an override for multiple aliases
-            // Use one of the module aliases specified for now
-            if pkgToAliasesMap[elemPkg]?[keyName] == nil, !elemAlias.isEmpty {
-                pkgToAliasesMap[elemPkg] = [keyName: elemAlias]
+        
+        if result.isEmpty {
+            return
```

Comment:
Is this correct in all cases? If we didn't get any module aliases so far, there might still be some in the recursive dependency tree, right?